### ISSUE TITLE
Redis 2.8.9 crashes on WinServer 2012 R2 with VirtualAlloc/COWAlloc fail

### ISSUE THREAD

**Issue by paveliak** [microsoftarchive/redis#133]
I have no problems running Redis 2.8.9 on Windows 8.1. However on WinServer 2012 R2 it crashes with VirtualAlloc/COWAlloc fail.

It crashes when memory limit is reached. So if I set "maxheap 1Mb" it won't even start. If I set it to "maxheap 10Mb" then it starts but crashes when client tries to execute a command. When I set it to 1Gb it runs fine until memory limit is reached and Redis is supposed to evict keys. It perfectly evicts them on Windows 8.1 freeing up memory for the new keys but crashes on Server 2012 R2.

I tried to rebuild sources with COW disabled and then all problems are gone however there are no memory limits enforced with VirtualAlloc.

**Comment by jepickett** [comment]
Can you send me your conf file?

**Comment by jepickett** [comment]
My guess here is that you are using the 'maxheap' flag without the 'maxmemory' flag. 'maxheap' by itself will not cause eviction to happen. If 'maxmemory' is used, then eviction will happen when a new key is created and there is already more than the 'maxmemory' amount of data allocated. 

The amount of allocated memory and the size of the heap are not the same. Redis allocations live in an unmanaged heap (it is written in C/C++, rather than a heap managed language like C# or Java). Thus there is no possibility of moving allocations around when memory pressure is high. Fragmentation is guaranteed. Thus the heap must be larger than the amount of memory you are going to allocate. By default the 'maxmemory' flag will set the 'maxheap' flag to be 50% larger than the 'maxmemory' setting. If fragmentation is very high, then this 50% head room may not be enough. Having Redis exit with an OOM error is an indication that you may have excessive fragmentation. The way to work around this is to specify a 'maxheap' value in addition to the 'maxmemory' value, and to make the 'maxheap' value significantly larger than the 50% headroom given by the 'maxmemory' flag by default.

You should also be aware that eviction will only occur on new key creation. If you have an instance of Redis with a list that is continuously appended to, you will eventually run out of memory without any eviction.

**Comment by paveliak** [comment]
Hi Jonathan

Sorry for the late reply. You are right, I was using "maxheap" without "maxmemory". What puzzles me is that such configuration crashes. However specifying "maxmemory" only or both "maxheap" and "maxmemory" works fine.

My configuration file is pretty simple. I just take default redis.windows.conf shipped with redis-2.8.9.zip and add these lines there:
save ""
#maxmemory 10Mb
maxheap 10Mb

Then I fill cache with this program (using StackExchange.Redis)

```
        var conn = ConnectionMultiplexer.Connect("localhost");
        for (int i = 0;  ;i++)
        {
            conn.GetDatabase().StringSet(i.ToString(), new string('X', 1024), TimeSpan.FromHours(1));
        }
```

It would crash if "maxheap" is specified without "maxmemory". And would work fine evicting the keys if both switches or "maxmemory" only is specified.

Thanks for heads up on key eviction. This is good to know.

**Comment by jepickett** [comment]
maxmemory is a flag that exists in the POSIX version of Redis for configuring eviction. Because we are trying to be as faithful to the original as we can, we are not touching the behavior of that flag. In the Windows port we had to add maxheap for our fork() simulation code to allow for heap configuration in more complicated scenarios. I do not expect that most users will ever have to use this flag. maxmemory with our silent configuration of maxheap is usually enough.
